### PR TITLE
✨ push service status to frontend

### PR DESCRIPTION
**Pull request by GitHK**
## What do these changes do?

- ♻️ Changes the start service interface which now requires the `primary_group_id` of the user. This one is used by socketio to deliver notifications to all open tabs of the user.
- ♻️ `dynamic-scheduler` uses the distributed resource identifier to keep track of started services.
- ✨ `dynamic-scheduler` delivers service state changes via socket.io to the frontend.

## Related issue/s

- https://github.com/ITISFoundation/osparc-issues/issues/1175
- https://github.com/ITISFoundation/osparc-issues/issues/1170

## Dev Checklist

- [x] No ENV changes or I properly updated ENV ([read the instruction](https://git.speag.com/oSparc/osparc-ops-deployment-configuration/-/blob/configs/README.md?ref_type=heads#how-to-update-env-variables))

## DevOps Checklist

- ~~⚠️ changed existing env var for ops stack see file `services/docker-compose-ops.yml` for details. Relative PR in ops https://github.com/ITISFoundation/osparc-ops-environments/pull/501~~ env var changes are already in master

### CODE REVIEW

File: services/docker-compose-ops.yml [hunk 1]
```diff
@@ -82,7 +82,8 @@ services:
         scheduled_maintenance:${REDIS_HOST}:${REDIS_PORT}:3,
         user_notifications:${REDIS_HOST}:${REDIS_PORT}:4,
         announcements:${REDIS_HOST}:${REDIS_PORT}:5,
-        distributed_identifiers:${REDIS_HOST}:${REDIS_PORT}:6
+        distributed_identifiers:${REDIS_HOST}:${REDIS_PORT}:6,
+        distributed_caches:${REDIS_HOST}:${REDIS_PORT}:7
```
Comment (GitHK) [inline services/docker-compose-ops.yml:86]: @YuryHrytsuk I'm not sure if the relative env var needs to be updated somewhere in OPS to reflect these changes

**Comment by mrnicegyu11** [comment]
tl;dr: I need more context.

@GitHK Sorry before I review:
This has >1.3k line changes, to understand this I please need some more explanations what and why this PR does what it does. 

My questions are specifically:
- I dot see frontend changes, so I dont get the PR's title :(
- `Changes the start service interface which now requires the primary_group_id of the user. This one is used by socketio to deliver notifications to all open tabs of the user.` Sentence one and two don't relate to each other on my mind... and I dont get why the `primary_group_id` needs to be included now. The linked issues are also not written very clearly, I dont understand them without more context
- I didnt spend too much time yet since there are too many changes for my mind, but on first site I think this PR needs to be accompanied by changes in the osparc-ops repo, since you change the CLI flags in the call to redis. Be careful when you do those things, I think merging as-is will cause trouble! There might be more, not sure at this point.



Sorry!

**Comment by GitHK** [comment]
> tl;dr: I need more context.
> 
> @GitHK Sorry before I review: This has >1.3k line changes, to understand this I please need some more explanations what and why this PR does what it does.
> 

@mrnicegyu11 you got dragged into this because you are a code owner. Also @YuryHrytsuk got dragged into this for the same reason. I would ask help form you both on the topics related to OPS.

Most likely you are interested in these two changes two files shown in the screenshot below. I require a hand with those. Please note my comment/question in the `DevOps Checklist` above. I need to understand if there are changes required to the OPS part.

Screenshot 2024-01-10 at 09 01 50


--- 

> My questions are specifically:
> 
> * I dot see frontend changes, so I dont get the PR's title :(

the backend already has all the logic in place to push notifications via socket.io to the frontend, that's why you see no changes to the JavaScript code.

> * `Changes the start service interface which now requires the primary_group_id of the user. This one is used by socketio to deliver notifications to all open tabs of the user.` Sentence one and two don't relate to each other on my mind... and I dont get why the `primary_group_id` needs to be included now. The linked issues are also not written very clearly, I dont understand them without more context

You were the wrong audience for the message. That is a backend implementation detail. The socketio messages are sent to users via their groups ids.

> * I didnt spend too much time yet since there are too many changes for my mind, but on first site I think this PR needs to be accompanied by changes in the osparc-ops repo, since you change the CLI flags in the call to redis. Be careful when you do those things, I think merging as-is will cause trouble! There might be more, not sure at this point.

We did this in the past, I think I had some assistance from @YuryHrytsuk (I have already tagged him in a question about it)

> 
> Sorry!

**Comment by mrnicegyu11** [comment]
Thanks a lot @GitHK that makes it clear :--)

File: services/docker-compose-ops.yml (hunk 1 above)
Comment (mrnicegyu11) [reply services/docker-compose-ops.yml:86]: It does!

Comment (mrnicegyu11) [reply services/docker-compose-ops.yml:86]: The new DB needs to be added here in this line https://github.com/ITISFoundation/osparc-ops-environments/blob/9c79757dfd9ff13ddbcb277cc44ddadab770e5e7/services/redis-commander/docker-compose.yml#L12

File: services/docker-compose.yml [hunk 2]
```diff
@@ -1030,7 +1030,7 @@ services:
         "--loglevel",
         "verbose",
         "--databases",
-        "7",
+        "8",
```
Comment (mrnicegyu11) [inline services/docker-compose.yml:1033]: see my comment above, w.r.t. the needed change in redis-commander :--)

**Review by mrnicegyu11 — CHANGES_REQUESTED** [review]
This needs a sister-PR in ops, with a line change here https://github.com/ITISFoundation/osparc-ops-environments/blob/9c79757dfd9ff13ddbcb277cc44ddadab770e5e7/services/redis-commander/docker-compose.yml#L12 adding the new Redis DB

:--)

File: services/docker-compose-ops.yml (hunk 1 above)
Comment (GitHK) [reply services/docker-compose-ops.yml:86]: thanks @mrnicegyu11 I have a PR in place now. https://github.com/ITISFoundation/osparc-ops-environments/pull/501
Also please let me know when this new PR should be merged in.

File: packages/service-library/src/servicelib/rabbitmq/rpc_interfaces/dynamic_scheduler/services.py [hunk 3]
```diff
@@ -65,6 +67,7 @@ async def stop_dynamic_service(
         node_id=node_id,
         simcore_user_agent=simcore_user_agent,
         save_state=save_state,
+        primary_group_id=primary_group_id,
```
Comment (sanderegg) [inline packages/service-library/src/servicelib/rabbitmq/rpc_interfaces/dynamic_scheduler/services.py:70]: side-comment: I think parsing the name of the method (line 66) every single time this gets called is an overkill. either create a constant and parse it only once, or just cast it, we know it passes the conventions.

File: packages/service-library/src/servicelib/base_distributed_identifier.py [hunk 4]
```diff
@@ -66,10 +70,14 @@ def __init__(
         self._cleanup_task: Task | None = None
 
     async def setup(self) -> None:
+        lock_key = f"lock:{self.class_path()}:base-distributed-identifier"
+        lock_value = json.dumps({})
         self._cleanup_task = start_periodic_task(
-            self._cleanup_unused_identifiers,
+            exclusive(self._redis_client_sdk, lock_key=lock_key, lock_value=lock_value)(
```
Comment (sanderegg) [inline packages/service-library/src/servicelib/base_distributed_identifier.py:76]: regarding your comment in the mattermost chat. I think you misunderstood the idea of using exclusive here. this effectively means that _cleanup_unused_identifiers will be run by only 1 process at a time. not sure what you were looking for here.

File: services/autoscaling/tests/manual/docker-compose.yml [hunk 5]
```diff
@@ -48,7 +48,8 @@ services:
         scheduled_maintenance:${REDIS_HOST}:${REDIS_PORT}:3,
         user_notifications:${REDIS_HOST}:${REDIS_PORT}:4,
         announcements:${REDIS_HOST}:${REDIS_PORT}:5,
-        distributed_identifiers:${REDIS_HOST}:${REDIS_PORT}:6
+        distributed_identifiers:${REDIS_HOST}:${REDIS_PORT}:6,
+        distributed_caches:${REDIS_HOST}:${REDIS_PORT}:7
```
Comment (sanderegg) [inline services/autoscaling/tests/manual/docker-compose.yml:52]: please revert that one. it is unnecessary and I will remove that stuff anyway from there. autoscaling uses redis in a very simple manner and does not need this. thanks and sorry for the confusion

File: services/clusters-keeper/docker-compose.yml [hunk 6]
```diff
@@ -43,7 +43,8 @@ services:
         scheduled_maintenance:${REDIS_HOST}:${REDIS_PORT}:3,
         user_notifications:${REDIS_HOST}:${REDIS_PORT}:4,
         announcements:${REDIS_HOST}:${REDIS_PORT}:5,
-        distributed_identifiers:${REDIS_HOST}:${REDIS_PORT}:6
+        distributed_identifiers:${REDIS_HOST}:${REDIS_PORT}:6,
+        distributed_caches:${REDIS_HOST}:${REDIS_PORT}:7
```
Comment (sanderegg) [inline services/clusters-keeper/docker-compose.yml:47]: and I think that is the same here. will have to review this anyway. thanks

File: services/dynamic-scheduler/requirements/_base.in [hunk 7]
```diff
@@ -14,9 +14,10 @@
 --requirement ../../../packages/service-library/requirements/_fastapi.in
 
 
-
+aiocache[redis]
 fastapi
 httpx
 packaging
+python-socketio
```
Comment (sanderegg) [inline services/dynamic-scheduler/requirements/_base.in:21]: I think there is an installation mode to "only" install the necessary stuff for sending messages instead of the whole package. maybe ask @pcrespov for this.

File: services/dynamic-scheduler/requirements/_test.in [hunk 8]
```diff
@@ -16,6 +16,7 @@ coverage
 docker
 faker
 pytest
+pytest-aiohttp
```
Comment (sanderegg) [inline services/dynamic-scheduler/requirements/_test.in:19]: as far as I know, that package is replaced by pytest-asyncio completely and was deprecated. Please check that and remove it.

File: services/dynamic-scheduler/src/simcore_service_dynamic_scheduler/services/services_tracker/_status_cache.py [hunk 9]
```diff
@@ -0,0 +1,35 @@
...
+
+    async def set_value(self, key: str, value: T) -> None:
+        await self.cache.set(key, value)  # pylint:disable=no-member
+
+    async def get_value(self, key: str) -> T | None:
+        result: T | None = await self.cache.get(key)  # pylint:disable=no-member
+        return result
+
+    async def clear(self) -> None:
+        await self.cache.clear()  # pylint:disable=no-member
```
Comment (sanderegg) [inline services/dynamic-scheduler/src/simcore_service_dynamic_scheduler/services/services_tracker/_status_cache.py:35]: I actually always wonder about the usefulness of creating functions that directly call another function. anyway.
why not using a dataclass where you declare the type and then you do not need all that pylint fidling??

File: services/dynamic-scheduler/src/simcore_service_dynamic_scheduler/services/services_tracker/_tracker.py [hunk 10]
```diff
@@ -0,0 +1,85 @@
...
+    async def setup(self) -> None:
+        await super().setup()
+
+        lock_key = f"lock:{self.class_path()}:services-tracker"
+        lock_value = json.dumps({})
+        self._check_task = start_periodic_task(
+            exclusive(self._redis_client_sdk, lock_key=lock_key, lock_value=lock_value)(
+                self._check_services_status_task
+            ),
+            interval=self.cleanup_interval,
```
Comment (sanderegg) [inline services/dynamic-scheduler/src/simcore_service_dynamic_scheduler/services/services_tracker/_tracker.py:48]: so you use that pattern here again. that means only one instance of dynamic scheduler can run the task at a time. but that does not prevent another of running it too.
Now it depends on what you wanna do...

File: services/dynamic-scheduler/tests/unit/api_rest/test_api_rest__health.py [hunk 11]
```diff
@@ -11,7 +11,9 @@
 
 @pytest.fixture
 def app_environment(
-    disable_rabbitmq_setup: None, app_environment: EnvVarsDict
+    disable_services_tracker_setup: None,
+    disable_rabbitmq_setup: None,
+    app_environment: EnvVarsDict,
```
Comment (sanderegg) [inline services/dynamic-scheduler/tests/unit/api_rest/test_api_rest__health.py:16]: question: do you actually have a real healthcheck or it only always returns 200?
in case redis is kaputt? or in case rabbitmq is kaputt?

File: services/dynamic-scheduler/tests/unit/api_rpc/test_api_rpc__services.py [hunk 12]
```diff
@@ -354,6 +358,7 @@ async def test_stop_dynamic_service(
         node_id=node_id,
         simcore_user_agent=simcore_user_agent,
         save_state=save_state,
+        primary_group_id=1,
```
Comment (sanderegg) [inline services/dynamic-scheduler/tests/unit/api_rpc/test_api_rpc__services.py:361]: the groupID 1 is usually the everyone group. I would try to not use it in tests. you might get tricked by that fact...
Also, I think we might have fixtures for creating group ids...

File: services/dynamic-scheduler/tests/unit/test_services_services_tracker__notfier.py [hunk 13]
```diff
@@ -0,0 +1,261 @@
...
+@pytest.fixture
+async def services_tracker(app: FastAPI) -> AsyncIterable[ServicesTracker]:
+    tracker = get_services_tracker(app)
+    yield tracker
+    # cleanup distributed cache between tests to avoid flakiness
+    await tracker.service_status_cache.clear()
+
+
+@pytest.fixture
+async def socketio_server(app: FastAPI) -> AsyncIterable[AsyncServer]:
```
Comment (sanderegg) [inline services/dynamic-scheduler/tests/unit/test_services_services_tracker__notfier.py:68]: check if not avaiable in pytest-simcore. and if not consider moving this there. I think @pcrespov also uses it this way.

File: services/dynamic-scheduler/tests/unit/test_services_services_tracker__notfier.py [hunk 14]
```diff
@@ -0,0 +1,261 @@
...
+@pytest.fixture
+async def web_server(
+    socketio_server: AsyncServer, aiohttp_unused_port: Callable
+) -> AsyncIterator[URL]:
+    """
+    this emulates the webserver setup: socketio server with
+    an aiopika manager that attaches an aiohttp web app
+    """
+    aiohttp_app = web.Application()
+    socketio_server.attach(aiohttp_app)
```
Comment (sanderegg) [inline services/dynamic-scheduler/tests/unit/test_services_services_tracker__notfier.py:93]: question: why do you care to simulate the webserver here? these are unit tests of the dynamic scheduler right?
in case you want to check the webserver, then do an integration test. otherwise I would not complicate stuff by adding aiohttp stuff here.

**Review by sanderegg — CHANGES_REQUESTED** [review]
this is looking promising. Please have a look at my comments, we can discuss them later. Also please revert the changes regarding autoscaling tests and clusters-keeper thanks.

**Review by YuryHrytsuk — APPROVED** [review]

**Review by pcrespov — COMMENTED** [review]
Sorry I have just noticed I ddi not press "submit"and the reviewed stayed there for a while...

**Review by matusdrobuliak66 — COMMENTED** [review]
There are a lot of reviews, I will wait after revision.

File: services/docker-compose.yml (hunk 2 above)
Comment (GitHK) [reply services/docker-compose.yml:1041]: OK the PR should be all set in OPS https://github.com/ITISFoundation/osparc-ops-environments/pull/501

Comment (GitHK) [reply services/docker-compose.yml:1041]: also the bump from 7 to 8 should only be required here

**Review by jsaq007 — COMMENTED** [review]
👍🏼

**Review by jsaq007 — APPROVED** [review]

**Comment by GitHK** [comment]
outdated closing in favour of new one
